client/client.go

```diff
@@ -99,6 +99,14 @@ type OpenstackClient struct {
 	controllerID string
 }
 
+func SetOpenstackClient(client *OpenstackClient, compute, image, network, volume *gophercloud.ServiceClient, controllerID string) {
```

Comment:
I think this can be a `NewTestOpenStackClient()` that creates a proper testing client instead of replacing individual service clients inside the already instantiated one.

You can then replace the old one with the one returned here:

```go
func NewTestOpenStackClient(mockClient *gophercloud.ServiceClient, controllerID string) *OpenstackClient {
    return &OpenstackClient{
		compute:      mockClient,
		image:        mockClient,
		network:      mockClient,
		volume:       mockClient,
		controllerID: controllerID,
	}
}
```

And then later in your code, you can just:

```go
serviceClient := thclient.ServiceClient()
mockCli := NewTestOpenStackClient(serviceClient, "my-controller-id")
provider.cli = mockCli
```

Reply:
This is indeed better! Updated it!